executor/ddl_test.go

```diff
@@ -676,6 +683,92 @@ func (s *testSuite8) TestShardRowIDBits(c *C) {
 	c.Assert(autoid.ErrAutoincReadFailed.Equal(err), IsTrue, Commentf("err:%v", err))
 }
 
+type testAutoRandomSuite struct {
+	*baseTestSuite
+}
+
+func (s *testAutoRandomSuite) TestAutoRandomBitsData(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+
+	tk.MustExec("create database if not exists test_auto_random_bits")
+	defer tk.MustExec("drop database if exists test_auto_random_bits")
+	tk.MustExec("use test_auto_random_bits")
+	tk.MustExec("drop table if exists t")
+
+	config.AutoRandom.SetupTestConfig()
+	defer config.AutoRandom.RestoreTestConfig()
+
+	tk.MustExec("create table t (a bigint primary key auto_random(15), b int)")
+	insertingValues := make([]string, 100)
+	for i := 0; i < 100; i++ {
+		insertingValues[i] = fmt.Sprintf("(%d)", i)
+	}
+	tk.MustExec(fmt.Sprintf("insert into t (b) values %s;", strings.Join(insertingValues, ",")))
+
+	dom := domain.GetDomain(tk.Se)
+	tbl, err := dom.InfoSchema().TableByName(model.NewCIStr("test_auto_random_bits"), model.NewCIStr("t"))
+	c.Assert(err, IsNil)
+	c.Assert(tk.Se.NewTxn(context.Background()), IsNil)
+	var allHandles []int64
+	// Iterate all the record. The order is not guaranteed.
+	err = tbl.IterRecords(tk.Se, tbl.FirstKey(), nil, func(h int64, _ []types.Datum, _ []*table.Column) (more bool, err error) {
+		allHandles = append(allHandles, h)
+		return true, nil
+	})
+	c.Assert(err, IsNil)
+	tk.MustExec("drop table t")
+
+	// Test auto random id number.
+	c.Assert(len(allHandles), Equals, 100)
+	// Test the first 15 bits of each handle is greater than 0.
+	for _, h := range allHandles {
+		c.Assert(h>>(64-15), Greater, int64(0))
+	}
+	// Test auto random id is monotonic increasing and continuous.
+	orderedHandles := make([]int64, len(allHandles))
+	for i, h := range allHandles {
+		orderedHandles[i] = h << 16 >> 16
+	}
+	sort.Slice(orderedHandles, func(i, j int) bool { return orderedHandles[i] < orderedHandles[j] })
+	size := int64(len(allHandles))
+	for i := int64(1); i <= size; i++ {
+		c.Assert(i, Equals, orderedHandles[i-1])
+	}
+
+	// Test explicit insert.
+	tk.MustExec("create table t (a tinyint primary key auto_random(2), b int)")
+	insertingValues = make([]string, 100)
+	for i := 0; i < 100; i++ {
+		insertingValues[i] = fmt.Sprintf("(%d, %d)", i, i)
+	}
+	tk.MustExec(fmt.Sprintf("insert into t values %s;", strings.Join(insertingValues, ",")))
```

Comment:
Explicit insert will succeed when the value > 31, but implicit insert will check it. is that right?

Reply:
Yes, even if the automatically allocated values are exhausted, explicit insertion is still successful.